File: 383-ransom-note/ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        char_count = {}

        for char in magazine:
            char_count[char] = char_count.get(char, 0) + 1

        for char in ransomNote:
            if char_count.get(char, 0) <= 0:
                return False
            char_count[char] -= 1

        return True
    

File: 383-ransom-note/ransom-note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function (ransomNote, magazine) {
  const charCount = {};

  // Count characters in magazine
  for (const char of magazine) {
    charCount[char] = (charCount[char] || 0) + 1;
  }

  // Check if characters in ransomNote can be constructed
  for (const char of ransomNote) {
    if (charCount[char]) {
      charCount[char]--;
    } else {
      return false;
    }
  }

  return true;
};
